src/include/migraphx/op/pooling.hpp

```diff
@@ -89,6 +92,135 @@ struct pooling
         check_attribute_size();
         return stride.size();
     }
+
+    struct lpnorm_pool
+    {
+        int p = 0;
+
+        lpnorm_pool() = delete;
+
+        explicit lpnorm_pool(int x) : p{x} {};
+
+        template <class T>
+        double init()
+        {
+            return 0.0;
+        }
+
+        double operator()(double x, double y) const { return x + std::pow(std::abs(y), p); }
+
+        double final(double x, std::size_t) const { return std::pow(x, 1. / p); }
+    };
+
+    struct avg_pool
+    {
+
+        template <class T>
+        double init()
+        {
+            return 0.0;
+        }
+
+        double operator()(double x, double y) { return x + y; }
+
+        double final(double x, std::size_t y) { return (y == 0) ? 0.0 : (x / y); }
+    };
+
+    struct max_pool
+    {
+        template <class T>
+        T init()
+        {
+            return std::numeric_limits<T>::lowest();
+        }
+
+        double operator()(double x, double y) { return std::max(x, y); }
+
+        double final(double x, std::size_t) { return (x); }
+    };
+
+    template <typename TOut, typename TIn, typename Op>
+    double calc_pooling(const TIn& input,
```

Comment:
The `calc_pooling` should contain the entire `par_for` loop.

Reply:
This will reduce the number of parameters needed to pass. It should only need the input, output tensor and the `Op`. Everything else can be calculated in this function before `par_for`.